integration/network/network.go

```diff
@@ -0,0 +1,173 @@
+	AzureCalicoSubnetMask = 17
+	AzureMasterSubnetMask = 24
+	AzureWorkerSubnetMask = 24
+)
+
+func init() {
+	var err error
+
+	var azureNetwork *net.IPNet
+	{
+		azureNetwork, err = AzureNetwork()
+		if err != nil {
+			panic(err)
+		}
+		os.Setenv(EnvAzureCIDR, azureNetwork.String())
+	}
+
+	var azureMasterSubnet *net.IPNet
+	{
+		azureMasterSubnet, err = AzureSubnet(EnvAzureMasterSubnetCIDR, *azureNetwork, AzureMasterSubnetMask)
+		if err != nil {
+			panic(err)
+		}
+		os.Setenv(EnvAzureMasterSubnetCIDR, azureMasterSubnet.String())
+	}
+
+	var azureWorkerSubnet *net.IPNet
+	{
+		azureWorkerSubnet, err = AzureSubnet(EnvAzureWorkerSubnetCIDR, *azureNetwork, AzureWorkerSubnetMask, *azureMasterSubnet)
+		if err != nil {
+			panic(err)
+		}
+		os.Setenv(EnvAzureWorkerSubnetCIDR, azureWorkerSubnet.String())
+	}
+
+	var azureCalicoSubnet *net.IPNet
+	{
+		azureCalicoSubnet, err = AzureSubnet(EnvAzureCalicoSubnetCIDR, *azureNetwork, AzureCalicoSubnetMask, *azureMasterSubnet, *azureWorkerSubnet)
+		if err != nil {
+			panic(err)
+		}
+		os.Setenv(EnvAzureCalicoSubnetCIDR, azureCalicoSubnet.String())
+	}
+}
+
+// AzureNetwork either return network from CIDR found at EnvAzureCIDR environement variable or
+// determine network using EnvE2ECIDR, EnvE2EMask and EnvCircleBuildNumber.
+func AzureNetwork() (azureNetwork *net.IPNet, err error) {
+	azureCDIR := os.Getenv(EnvAzureCIDR)
+	if azureCDIR == "" {
+		e2eCIDR := os.Getenv(EnvE2ECIDR)
+
+		e2eMask, err := strconv.ParseUint(os.Getenv(EnvE2EMask), 10, 32)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		buildNumber, err := strconv.ParseUint(os.Getenv(EnvCircleBuildNumber), 10, 32)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		azureNetwork, err = DetermineSubnet(e2eCIDR, uint(e2eMask), uint(buildNumber))
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	} else {
+		_, azureNetwork, err = net.ParseCIDR(azureCDIR)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return azureNetwork, nil
+}
+
+// AzureSubnet either return subnet from CIDR found at envVar environement variable or
+// compute the next free subnet using network, mask and allocatedSubnet (if any).
+func AzureSubnet(envVar string, network net.IPNet, mask int, allocatedSubnet ...net.IPNet) (subnet *net.IPNet, err error) {
+	subnetCIDR := os.Getenv(envVar)
+	if subnetCIDR == "" {
+		s, err := ipam.Free(network, net.CIDRMask(mask, 32), allocatedSubnet)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		subnet = &s
+	} else {
+		_, subnet, err = net.ParseCIDR(subnetCIDR)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return subnet, nil
+}
+
+// DetermineSubnet is deterministic, it will always return the same network given the same input.
+// It compute the number of available networks between cidr and mask.
+// And use `decider mod available networks` to determine which one to pick.
+// e.g. DetermineSubnet("10.255.0.0/9", 16, 42)  >  10.170.0.0/16
```

Comment:
`10.255.0.0/9` this isn't correct CIDR network notation (even though parsable by golang library) - it contains network + IP. I don't want to read the code below, because I don't understand how it is supposed to work. I guess it always changes the second octet. If so why not just use string formatting? That's way simpler to reason about.

Reply:
Yeah true, the correct CIDR here is `10.128.0.0/9` .

It is not always a whole octet it changes depending on the given mask and CIDR's mask. Let me explain with a full example
```
CIDR: 10.128.0.0/9    Mask: 17    Decider: 1

CIDR mask: 255.0.0.0 = 9        11111111.10000000.00000000.00000000
Mask: 255.255.0.0 = 17          11111111.11111111.10000000.00000000
                                          ^       ^ = 8 bits

Leaving us with (7 bits=) 256 possible subnets.
We wrap the decider in those possibilities: 1 mod 256 = 1

How much do we need to shift 1: 32 (bits in IP) - 17 (mask) = 15
1 bit shifted 15 times left = 10000000.00000000 > 0.0.128.0

We compute the final CIDR:
10.128.0.0 >    00001010.10000000.00000000.00000000
OR
0.0.128.0 >     00000000.00000000.10000000.00000000
                -----------------------------------
10.128.128.0 >  0001010.10000000.10000000.00000000
```
I am not sure there is a easier way to compute this.